types/src/v0/impls/header.rs

```diff
@@ -132,6 +117,276 @@ impl Committable for Header {
     }
 }
 
+impl Header {
+    pub fn version(&self) -> Version {
+        match self {
+            Self::V1(_) => Version { major: 0, minor: 1 },
+            Self::V2(_) => Version { major: 0, minor: 2 },
+            Self::V3(_) => Version { major: 0, minor: 3 },
+        }
+    }
+}
+
+impl From<Version> for ResolvableChainConfigOrVersion {
+    fn from(version: Version) -> Self {
+        Self {
+            chain_config: EitherOrVersion::Version(version),
+        }
+    }
+}
+
+impl From<ResolvableChainConfig> for ResolvableChainConfigOrVersion {
+    fn from(v: ResolvableChainConfig) -> Self {
+        let value = match v.chain_config {
+            Either::Left(cfg) => EitherOrVersion::Left(cfg),
+            Either::Right(commit) => EitherOrVersion::Right(commit),
+        };
+
+        Self {
+            chain_config: value,
+        }
+    }
+}
+
+impl Serialize for Header {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Self::V1(header) => header.serialize(serializer),
+            Self::V2(fields) => VersionedHeader {
+                version: Version { major: 0, minor: 2 }.into(),
+                fields: fields.clone(),
+            }
+            .serialize(serializer),
+            Self::V3(fields) => VersionedHeader {
+                version: Version { major: 0, minor: 3 }.into(),
+                fields: fields.clone(),
+            }
+            .serialize(serializer),
+        }
+    }
+}
+
+#[derive(Deserialize, Debug, PartialEq, Eq, Hash, strum::Display)]
+#[serde(field_identifier, rename_all = "snake_case")]
+#[strum(serialize_all = "snake_case")]
+pub enum StructFields {
+    Version,
+    Fields,
+    ChainConfig,
+    Height,
+    Timestamp,
+    L1Head,
+    L1Finalized,
+    PayloadCommitment,
+    BuilderCommitment,
+    NsTable,
+    BlockMerkleTreeRoot,
+    FeeMerkleTreeRoot,
+    FeeInfo,
+    BuilderSignature,
+}
+
+impl StructFields {
+    fn list() -> &'static [&'static str] {
+        &[
+            "fields",
+            "chain_config",
+            "height",
+            "timestamp",
+            "l1_head",
+            "l1_finalized",
+            "payload_commitment",
+            "builder_commitment",
+            "ns_table",
+            "block_merkle_tree_root",
+            "fee_merkle_tree_root",
+            "fee_info",
+            "builder_signature",
+        ]
+    }
+}
```

Comment:
We can remove the `StructField` but we still need hard coded list of fields which can be passed to `deserialize_struct()`

Reply:
Hmm...technically yes, but actually the list of fields we're passing in isn't even meaningful. These are the union of all fields from all versions of the struct, but we're never going to encounter a struct that has exactly these fields. Luckily, this works, because serde_json ignores the fields, while bincode only cares about the length, which is only an upper bound at that: the bincode deserializer will only consume the next field [when `next_element` is called](https://docs.rs/bincode/latest/src/bincode/de/mod.rs.html#313), which our visitor does the correct number of times. This wouldn't work if, say, the bincode deserializer eagerly consumed one element for each field.

So, in summary, I guess listing all the fields from all versions is...more correct than anything else? But we need some comments explaining exactly what's happening here at a low level for each of the formats we care about (JSON and bincode), because as a reader it's easy to get the wrong impression, and there's a nonzero chance that a future update to bincode or serde_json could break this and we'll need to rely on comments to figure out what happened